config_options.md
# Configuration Options

Some functionality of large_image is controlled through configuration parameters.  These can be read or set via python using functions in the `large_image.config` module, [getConfig](./_build/large_image/large_image.html#large_image.config.getConfig) and [setConfig](./_build/large_image/large_image.html#large_image.config.setConfig).

#### Configuration Parameters

| Key(s)                                                                                                                                               | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Type                                                     | Default<br/><br/><a id="config-logger"></a>                                                                                                                                                                                                                                                                     |
|------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `logger` [🔗](#config-logger)                                                                                           | Most log messages are sent here.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `logging.Logger`                                         | This defaults to the standard python logger using the name large_image. When using Girder, this default to Girder’s logger, which allows colored console output.<br/><br/><a id="config-logprint"></a>                                                                                                          |
| `logprint` [🔗](#config-logprint)                                                                                         | Messages about available tilesources are sent here.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | `logging.Logger`                                         | This defaults to the standard python logger using the name large_image. When using Girder, this default to Girder’s logger, which allows colored console output.<br/><br/><a id="config-default-encoding"></a>                                                                                                  |
| `default_encoding` [🔗](#config-default-encoding)                                                                                 | Default encoding for tile sources.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | `'JPEG' | 'PNG' | 'TIFF' | 'TILED'`                      | `None`<br/><br/><a id="config-default-projection"></a>                                                                                                                                                                                                                                                          |
| `default_projection` [🔗](#config-default-projection)                                                                               | Default projection for geospatial tile sources. Use a proj4 projection string or a case-insensitive string of the form ‘EPSG:<epsg number>’.                                                                                                                                                                                                                                                                                                                                                                                                              | `str | None`                                             | `None`<br/><br/><a id="config-cache-backend"></a>                                                                                                                                                                                                                                                               |
| `cache_backend` [🔗](#config-cache-backend)                                                                                    | String specifying how tiles are cached.  If memcached is not available for any reason, the python cache is used instead.                                                                                                                                                                                                                                                                                                                                                                                                                                  | `None | str: "python" | str: "memcached" | str: "redis"` | `None` (When None, the first cache available in the order memcached, redis, python is used. Otherwise, the specified cache is used if available, falling back to python if not.)<br/><br/><a id="config-cache-python-memory-portion"></a>                                                                       |
| `cache_python_memory_portion` [🔗](#config-cache-python-memory-portion)                                                                      | If tiles are cached with python, the cache is sized so that it is expected to use less than 1 / (`cache_python_memory_portion`) of the available memory.                                                                                                                                                                                                                                                                                                                                                                                                  | `int`                                                    | `16`<br/><br/><a id="config-cache-memcached-url"></a>                                                                                                                                                                                                                                                           |
| `cache_memcached_url` [🔗](#config-cache-memcached-url)                                                                              | If tiles are cached in memcached, the url or list of urls where the memcached server is located.                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `str | List[str]`                                        | `"127.0.0.1"`<br/><br/><a id="config-cache-memcached-username"></a>                                                                                                                                                                                                                                             |
| `cache_memcached_username` [🔗](#config-cache-memcached-username)                                                                         | A username for the memcached server.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `str`                                                    | `None`<br/><br/><a id="config-cache-memcached-password"></a>                                                                                                                                                                                                                                                    |
| `cache_memcached_password` [🔗](#config-cache-memcached-password)                                                                         | A password for the memcached server.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `str`                                                    | `None`<br/><br/><a id="config-cache-redis-url"></a>                                                                                                                                                                                                                                                             |
| `cache_redis_url` [🔗](#config-cache-redis-url)                                                                                  | If tiles are cached in redis, the url or list of urls where the redis server is located.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  | `str | List[str]`                                        | `"127.0.0.1:6379"`<br/><br/><a id="config-cache-redis-username"></a>                                                                                                                                                                                                                                            |
| `cache_redis_username` [🔗](#config-cache-redis-username)                                                                             | A username for the redis server.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `str`                                                    | `None`<br/><br/><a id="config-cache-redis-password"></a>                                                                                                                                                                                                                                                        |
| `cache_redis_password` [🔗](#config-cache-redis-password)                                                                             | A password for the redis server.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `str`                                                    | `None`<br/><br/><a id="config-cache-tilesource-memory-portion"></a>                                                                                                                                                                                                                                             |
| `cache_tilesource_memory_portion` [🔗](#config-cache-tilesource-memory-portion)                                                                  | Tilesources are cached on open so that subsequent accesses can be faster.  These use file handles and memory.  This limits the maximum based on a memory estimation and using no more than 1 / (`cache_tilesource_memory_portion`) of the available memory.                                                                                                                                                                                                                                                                                               | `int`                                                    | `32` Memory usage by tile source is necessarily a rough estimate, since it can vary due to a wide variety of image-specific and deployment-specific details; this is intended to be conservative.<br/><br/><a id="config-cache-tilesource-maximum"></a>                                                         |
| `cache_tilesource_maximum` [🔗](#config-cache-tilesource-maximum)                                                                         | If this is zero, this signifies that `cache_tilesource_memory_portion` determines the number of sources that will be cached. If this greater than 0, the cache will be the smaller of the value computed for the memory portion and this value (but always at least 3).                                                                                                                                                                                                                                                                                   | `int`                                                    | `0`<br/><br/><a id="config-cache-sources"></a>                                                                                                                                                                                                                                                                  |
| `cache_sources` [🔗](#config-cache-sources)                                                                                    | If set to False, the default will be to not cache tile sources.  This has substantial performance penalties if sources are used multiple times, so should only be set in singular dynamic environments such as experimental notebooks.                                                                                                                                                                                                                                                                                                                    | `bool`                                                   | `True`<br/><br/><a id="config-max-small-image-size"></a>                                                                                                                                                                                                                                                        |
| `max_small_image_size` [🔗](#config-max-small-image-size)                                                                             | The PIL tilesource is used for small images if they are no more than this many pixels along their maximum dimension.                                                                                                                                                                                                                                                                                                                                                                                                                                      | `int`                                                    | `4096`  Specifying values greater than this could reduce compatibility with tile use on some browsers. In general, `8192` is safe for all modern systems, and values greater than `16384` should not be specified if the image will be viewed in any browser.<br/><br/><a id="config-source-ignored-names"></a> |
| `source_bioformats_ignored_names`,<br/>`source_pil_ignored_names`,<br/>`source_vips_ignored_names` [🔗](#config-source-ignored-names) | Some tile sources can read some files that are better read by other tilesources.  Since reading these files is suboptimal, these tile sources have a setting that, by default, ignores files without extensions or with particular extensions.                                                                                                                                                                                                                                                                                                            | `str` (regular expression)                               | Sources have different default values; see each source for its default. For example, the vips source default is `r'(^[^.]*|\.(yml|yaml|json|png|svs|mrxs))$'`<br/><br/><a id="config-all-sources-ignored-names"></a>                                                                                            |
| `all_sources_ignored_names` [🔗](#config-all-sources-ignored-names)                                                                        | If a file matches the regular expression in this setting, it will only be opened by sources that explicitly match the extension or mimetype.  Some formats are composed of multiple files that can be read as either a small image or as a large image depending on the source; this prohibits all sources that don’t explicitly support the format.                                                                                                                                                                                                      | `str` (regular expression)                               | `'(\.mrxs|\.vsi)$'`<br/><br/><a id="config-icc-correction"></a>                                                                                                                                                                                                                                                 |
| `icc_correction` [🔗](#config-icc-correction)                                                                                   | If this is True or undefined, ICC color correction will be applied for tile sources that have ICC profile information.  If False, correction will not be applied.  If the style used to open a tilesource specifies ICC correction explicitly (on or off), then this setting is not used.  This may also be a string with one of the intents defined by the PIL.ImageCms.Intents enum.  `True` is the same as `perceptual`.                                                                                                                               | `bool | str: one of PIL.ImageCms.Intents`                | `True`<br/><br/><a id="config-force-gdal-vsis3"></a>                                                                                                                                                                                                                                                            |

CHANGELOG.md
# Change Log

## 1.35.6

### Improvements

- Add an optional layerHotkeys prop to FrameSelector ([#2126](../../pull/2126))
- Close cgroup reads deterministically ([#2128](../../pull/2128))

### Changes

- Update a type to work with mypy 2.4.0 ([#2127](../../pull/2127))

## 1.35.5

### Improvements

- Find public config files blocked by private ones ([#2120](../../pull/2120))

### Changes

- Updates to work with wsidicom 0.30.x ([#2124](../../pull/2124))

## 1.35.4

### Improvements

- Add FORCE_GDAL_VSIS3 configuration option ([#2111](../../pull/2111))

## 1.35.3

### Improvements

- Better detect available memory in containers ([#2115](../../pull/2115))
- Avoid unclosed sqlite db when opening zarr source ([#2117](../../pull/2117))

### Changes

- Avoid a recent release of a package ([#2116](../../pull/2116))

## 1.35.2

### Improvements

- Allow customizing unselected annotation dimming ([#2113](../../pull/2113))

## 1.35.1

### Improvements

- Add a few missing type annotations ([#2074](../../pull/2074))

### Changes

- Added more options to the TestTileSource ([#2102](../../pull/2102))
- Change a band-count check for a PIL image to getbands() instead of mode ([#2104](../../pull/2104))

### Bug Fixes

- Fix zarr sink multi-process concurrency for python 3.14 ([#2098](../../pull/2098), [#2100](../../pull/2100))
- Fix subtile slicing for certain empty layer interpolations ([#2101](../../pull/2101))

## 1.35.0

### Features

- Eager iterator ([#2072](../../pull/2072), [#2073](../../pull/2073))

### Changes

- Set GDAL_PAM_ENABLED to NO if it is not set to avoid creating aux.xml files ([#2093](../../pull/2093))

## 1.34.3

### Improvements

- In Girder, by default include Updated in lists; do not show controls in dialog lists ([#2078](../../pull/2078))

### Changes

- Set an expiry duration on redis cache entries ([#2092](../../pull/2092))

### Bug Fixes

- Fixed a typo in the word "description" in the multi-source and annotation schemas ([#2082](../../pull/2082))

## 1.34.2

### Improvements

- Handle openslide keys in internal metadata with poor names ([#2061](../../pull/2061))

### Bug Fixes

- Fix handling single file DICOM files in Girder ([#2066](../../pull/2066))
- DICOM tile reads were not thread safe ([#2067](../../pull/2067))
- Handle more Redis errors ([#2068](../../pull/2068))

## 1.34.1

### Improvements

- Allow passing redis protocol as part of the redis cache url ([#2060](../../pull/2060))

### Changes

- Mark for girder < 5 ([#2054](../../pull/2054))

### Bug Fixes

- Reading a small region from a large tile could be wrong ([#2049](../../pull/2049))
- Fix a type in the histogram signature ([#2051](../../pull/2051))
- Don't fail when trying to read files pandas can't handle for plots ([#2059](../../pull/2059))

## 1.34.0

### Improvements

- Add support for Python 3.14 ([#2022](../../pull/2022), [#2026](../../pull/2026))
- In Jupyter, set edge value of tilesource to transparent ([#2019](../../pull/2019))
- When reading tiff files via the tiff or openslide reader, stop reading sooner if they are corrupt ([#2037](../../pull/2037))
- Support two point warp correspondence in the multi-source ([#2039](../../pull/2039))

### Changes

- Refactored internal code to use contextlib.suppress where appropriate ([#2031](../../pull/2031))
- Drop support for Python 3.9 ([#2032](../../pull/2032))
- When converting images, associated images stored as jpegs will be YCbCr ([#2043](../../pull/2043))

### Bug Fixes

- When reading raw files via the PIL source, fix handling uint16 data ([#2030](../../pull/2030))
- When copying an item via girder, ensure permissions on annotations are granted to the owner of the destination folder ([#2033](../../pull/2033))
- The large_image/folder/{}/tiles endpoint didn't always handle partially complete items ([#2038](../../pull/2038))

## 1.33.5

### Improvements

- Added a "pattern" property for polygon annotations ([#2020](../../pull/2020))

### Changes

- Change delattr to del when it doesn't matter ([#2021](../../pull/2021))

## 1.33.4

### Improvements

- Remove locking in reading nd2 tiles ([#1977](../../pull/1977))
- Get a pixel from multiple frames in a single REST call ([#2002](../../pull/2002))
- Add another index to speed up deleting large images in Girder ([#2005](../../pull/2005))

### Changes

- Increase the detail threshold for large annotations ([#1995](../../pull/1995))
- Update to work with tol-colors >= 2 ([#2000](../../pull/2000))
- Load some sources earlier in Girder ([#2004](../../pull/2004))
- Increase number of annotation elements processed for metadata plots ([#2012](../../pull/2012))
- More listed mimetypes ([#2013](../../pull/2013))
- Make it easier to rebind annotation models ([#2014](../../pull/2014))

### Bug Fixes

- Fixed recursion in the create large images in a folder endpoint ([#1993](../../pull/1993))
- When overlaying many image annotations, some showed artifacts ([#1998](../../pull/1998))
- Fix annotation caching with different users ([#1999](../../pull/1999))
- Work around how bioformats handles DICOMs in Monochrome1 ([#1834](../../pull/1834))
- Guard a missing data field in pixelmap annotation actions ([#2010](../../pull/2010), [#2011](../../pull/2011))
- Fixed an issue emitting geojson annotations with rotated rectangles and ellipses ([#2015](../../pull/2015))

## 1.33.3

### Changes

- Handle a recent change in pylibtiff ([#1992](../../pull/1992))

## 1.33.2

### Bug Fixes

- Fix some raw bson handling ([#1991](../../pull/1991))

## 1.33.1

### Improvements

- Speed up annotation centroid queries ([#1981](../../pull/1981), [#1990](../../pull/1990))
- Improve large annotation load and display speed ([#1982](../../pull/1982))
- Denormalize some values in the annotation collection to support improving display speed ([#1984](../../pull/1984))
- Improve indices for annotationelement queries ([#1985](../../pull/1985))
- Use some raw bson handling to speed up annotationelement serialization ([#1986](../../pull/1986))

large_image/config.py
import contextlib
import functools
import json
import logging
import os
import pathlib
import re
from typing import Any, cast

from . import exceptions

try:
    import psutil
    HAS_PSUTIL = True
except ImportError:
    HAS_PSUTIL = False

# Default logger
fallbackLogger = logging.getLogger('large_image')
fallbackLogger.setLevel(logging.INFO)
fallbackLogHandler = logging.NullHandler()
fallbackLogHandler.setLevel(logging.NOTSET)
fallbackLogger.addHandler(fallbackLogHandler)


def _in_notebook() -> bool:
    """
    Try to detect if we are in an interactive notebook.

    :returns: True if we think we are in an interactive notebook.
    """
    try:
        return get_ipython() is not None  # type: ignore[name-defined]
    except NameError:
        return False


ConfigValues = {
    'logger': fallbackLogger,
    'logprint': fallbackLogger,

    # Encoding and Projection
    'default_encoding': 'PNG' if _in_notebook() else 'JPEG',
    'default_projection': 'EPSG:3857' if _in_notebook() else None,

    # For tiles
    'cache_backend': None,  # 'python', 'redis' or 'memcached'
    # 'python' cache can use 1/(val) of the available memory
    'cache_python_memory_portion': 32,
    # cache_memcached_url may be a list
    'cache_memcached_url': '127.0.0.1',
    'cache_memcached_username': None,
    'cache_memcached_password': None,
    'cache_redis_url': '127.0.0.1:6379',
    'cache_redis_password': None,

    # If set to False, the default will be to not cache tile sources.  This has
    # substantial performance penalties if sources are used multiple times, so
    # should only be set in singular dynamic environments such as experimental
    # notebooks.
    'cache_sources': not _in_notebook(),

    # Generally, these keys are the form of "cache_<cacheName>_<key>"

    # For tilesources.  These are also limited by available file handles.
    # 'python' cache can use 1/(val) of the available memory based on a very
    # rough estimate of the amount of memory used by a tilesource
    'cache_tilesource_memory_portion': 16,
    # If >0, this is the maximum number of tilesources that will be cached
    'cache_tilesource_maximum': 0,

    'max_small_image_size': 4096,

    # Should ICC color correction be applied by default
    'icc_correction': True,

    # The maximum size of an annotation file that will be ingested into girder
    # via direct load
    'max_annotation_input_file_length': 1 * 1024 ** 3 if not HAS_PSUTIL else max(
        1 * 1024 ** 3, psutil.virtual_memory().total // 16),

    # Any path that matches here will only be opened by a source that matches
    # extension or mime type.
    'all_sources_ignored_names': r'(\.mrxs|\.vsi)$',

    # If True, make_vsi uses /vsis3/ for http(s) URLs instead of /vsicurl/.
    # Useful with MinIO when AWS_S3_ENDPOINT (and related GDAL/AWS env vars)
    # are configured.
    'force_gdal_vsis3': False,
}


# We use lru_cache over cache because it can be cleared
@functools.lru_cache(maxsize=1000)
def getConfig(key: str | None = None,
              default: str | bool | int | logging.Logger | None = None) -> Any:
    """
    Get the config dictionary or a value from the cache config settings.

    :param key: if None, return the config dictionary.  Otherwise, return the
        value of the key if it is set or the default value if it is not.
    :param default: a value to return if a key is requested and not set.
    :returns: either the config dictionary or the value of a key.
    """
    if key is None:
        return ConfigValues
    envKey = f'LARGE_IMAGE_{key.replace(".", "_").upper()}'
    if envKey in os.environ:
        value = os.environ[envKey]
        if value == '__default__':
            return default
        with contextlib.suppress(ValueError):
            value = json.loads(value)
        return value
    return ConfigValues.get(key, default)


def getLogger(key: str | None = None,
              default: logging.Logger | None = None) -> logging.Logger:
    """
    Get a logger from the config.  Ensure that it is a valid logger.

    :param key: if None, return the 'logger'.
    :param default: a value to return if a key is requested and not set.
    :returns: a logger.
    """
    logger = cast(logging.Logger, getConfig(key or 'logger', default))
    if not isinstance(logger, logging.Logger):
        logger = fallbackLogger
    return logger


def setConfig(key: str, value: str | bool | int | logging.Logger | None) -> None:
    """
    Set a value in the config settings.

    :param key: the key to set.
    :param value: the value to store in the key.
    """
    curConfig = getConfig()
    if curConfig.get(key) is not value:
        curConfig[key] = value
        getConfig.cache_clear()


def _ignoreSourceNames(
        configKey: str, path: str | pathlib.Path, default: str | None = None) -> None:
    """
    Given a path, if it is an actual file and there is a setting
    "source_<configKey>_ignored_names", raise a TileSourceError if the path
    matches the ignore names setting regex in a case-insensitive search.

    :param configKey: key to use to fetch value from settings.
    :param path: the file path to check.
    :param default: a default ignore regex, or None for no default.
    """
    ignored_names = getConfig('source_%s_ignored_names' % configKey) or default
    if not ignored_names or not os.path.isfile(path):
        return
    if re.search(ignored_names, os.path.basename(path), flags=re.IGNORECASE):
        raise exceptions.TileSourceError('File will not be opened by %s reader' % configKey)


def cpu_count(logical: bool = True) -> int:
    """
    Get the usable CPU count.  If psutil is available, it is used, since it can
    determine the number of physical CPUS versus logical CPUs.  This returns
    the smaller of that value from psutil and the number of cpus allowed by the
    os scheduler, which means that for physical requests (logical=False), the
    returned value may be more the the number of physical cpus that are usable.

    :param logical: True to get the logical usable CPUs (which include
        hyperthreading).  False for the physical usable CPUs.
    :returns: the number of usable CPUs.
    """
    count = os.cpu_count() or 2
    with contextlib.suppress(AttributeError):
        count = min(count, len(os.sched_getaffinity(0)))
    if HAS_PSUTIL:
        count = min(count, psutil.cpu_count(logical) or count)
    return max(1, count)


def total_memory() -> int:
    """
    Get the total memory in the system.  If this is in a container, try to
    determine the memory available to the cgroup.

    :returns: the available memory in bytes, or 8 GB if unknown.
    """
    mem = 0
    if HAS_PSUTIL:
        mem = psutil.virtual_memory().total
    caps = {'/sys/fs/cgroup/memory/memory.limit_in_bytes', '/sys/fs/cgroup/memory.max'}
    for cap in caps:
        try:
            cgroup = int(pathlib.Path(cap).read_text().strip())
            if 1024 ** 3 <= cgroup < 1024 ** 4 and (mem is None or cgroup < mem):
                mem = cgroup
        except Exception:
            pass
    if mem:
        return mem
    return 8 * 1024 ** 3


def minimizeCaching(mode: str | None = None) -> None:
    """
    Set python cache sizes to very low values.

    :param mode: None for all caching, 'tile' for the tile cache, 'source' for
        the source cache.
    """
    if not mode or str(mode).lower().startswith('source'):
        setConfig('cache_tilesource_maximum', 1)
    if not mode or str(mode).lower().startswith('tile'):
        setConfig('cache_backend', 'python')
        setConfig('cache_python_memory_portion', 256)
